File: src/lib/program-turn.ts
/**
 * 프로그램 매매 턴 손익 표시 계산 (route의 OFF 동결 + 프론트 실시간 표시 공용).
 *
 * 기록은 파이썬(program_turn.py)이 원장에 한다. 여기서는 확정분(by_tag)에
 * 현재 보유분의 미실현(현재가 − 기준가)을 얹어 '지금 시점의 턴 손익'을 만든다.
 */

export type ProgramTurn = {
    id: string;
    capital: number;
    basis: Record<string, number>;
    by_tag: Record<string, number>;
    active_tag: string | null;
    /** 파이썬이 매도 확정 시점에 누적하는, 이 턴에서 실제로 낸 매매 비용. */
    fees_realized?: number;
};

export type ProgramPosition = { name: string; quantity: number; avg_price: number; tag?: string };

export type TurnResult = { pnl: number; byTag: Record<string, number> };

/**
 * OFF 시점에 config에 동결되는 직전 턴 결과(표시 전용).
 * pnl === null 이면 '측정 불가'(계산에 필요한 조회가 실패/지연) — 진짜 0원 턴과 반드시 구분해 그린다.
 * degraded가 있으면 by_tag는 비어 있다(부분 합계가 전체인 것처럼 보이지 않게).
 */
/**
 * 실계좌에서 사라져 손익을 계상하지 못한 청산분.
 *
 * 프로그램이 낸 매도만 realized_pnl에 누적되므로, 수동 청산은 원장에서 포지션만
 * 빠지고 손익은 어디에도 남지 않는다. 체결가를 우리가 모르기 때문에 pnl 필드가
 * 없다 — 매입원가(cost_basis)만 적고 '미정산'으로 둔다. 값을 지어내는 대신
 * 누락 사실을 드러내기 위한 타입이다.
 */
export type UnreconciledExit = {
    date: string;
    code: string;
    name: string;
    quantity: number;
    avg_price: number;
    cost_basis: number;
};

export type LastTurnResult = {
    id: string;
    ended_at: string;
    /** 턴이 열린 시각. 없으면 '언제부터'를 그릴 수 없다(구 기록은 없을 수 있다). */
    started_at?: string;
    sim: string | null;
    capital: number;
    pnl: number | null;
    /** 이 턴에 실제로 낸 매매 비용. 계산 불가였으면 null. */
    fees?: number | null;
    by_tag: Record<string, number>;
    degraded?: 'ledger_unavailable' | 'prices_unavailable' | 'timeout';
};

/**
 * 턴 손익 = 확정분(by_tag) + 보유분 미실현(각 종목의 tag에 귀속).
 * 시세를 못 구한 종목은 기여분 0으로 처리한다(기존 프로그램 평가손익 계산과 동일한 폴백).
 */
export function computeTurnPnl(
    turn: ProgramTurn | null | undefined,
    positions: Record<string, ProgramPosition>,
    prices: Record<string, number>,
): TurnResult {
    if (!turn || !turn.id) return { pnl: 0, byTag: {} };

    const byTag: Record<string, number> = { ...(turn.by_tag || {}) };
    const basis = turn.basis || {};

    for (const [code, pos] of Object.entries(positions || {})) {
        const px = Number(prices[code]) || 0;
        if (px <= 0) continue;
        // ||(falsy 폴백): 파이썬 new_turn()의 `or` 체인과 동일. basis에 0이 들어오면
        // ??는 0을 통과시켜 (px-0)*qty = 시가총액 전체가 턴 수익으로 계상된다.
        const b = Number(basis[code] || px);
        const tag = pos.tag || turn.active_tag || 'unknown';
        byTag[tag] = (byTag[tag] || 0) + (px - b) * pos.quantity;
    }

    const pnl = Object.values(byTag).reduce((s, v) => s + v, 0);
    return { pnl, byTag };
}

/**
 * 턴 기준가 = 보유 종목의 **매입 평단**. 파이썬 new_turn과 같은 규칙이다.
 *
 * ON 시점 시세로 리셋(MTM)하지 않는다 — ON 전부터 보유한 종목도 원래 매입가부터
 * 재야 KIS 종목별 ROI와 턴 손익이 정합한다. 예전에는 route가 현재가로,
 * 파이썬이 평단으로 잡아 파이썬 첫 런 전후로 같은 턴의 손익이 점프했다.
 *
 * avg_price가 0/누락이면 항목을 만들지 않는다. 0을 기준가로 넣으면
 * (현재가 - 0) * 수량, 즉 시가총액 전체가 턴 수익으로 계상된다.
 */
export function basisFromPositions(
    positions: Record<string, ProgramPosition>,
): Record<string, number> {
    const basis: Record<string, number> = {};
    for (const [code, pos] of Object.entries(positions || {})) {
        const avg = Number(pos?.avg_price) || 0;
        if (avg > 0) basis[code] = avg;
    }
    return basis;
}

File: src/pipeline/workers/program_turn.py
"""프로그램 매매 턴 회계 (표시 전용, 순수 함수)
=======================================================
턴 = 프로그램 매매를 켠 시점부터 끈 시점까지.

기존 원장의 realized_pnl은 평단(avg_price) 기준으로 누적되며 effective_budget(복리)의
근거다 — 여기에는 절대 손대지 않는다. 턴 회계는 그와 분리된 별도 트랙으로, 기준가(basis)
기준으로 집계한다. 기준가는 턴 시작 시점에 **매입 평단**으로 seed된다(ON 시점 시세로
리셋하지 않는다). 전략 스위칭 시점에는 직전 태그 구간을 락인하며 기준가를 그 순간 시세로
리셋한다 — 하지만 seed가 평단이므로 구간 손익의 합은 평단 기준으로 telescoping된다.

이렇게 하면 ON 전부터 보유한 종목도 원래 매입가부터 손익이 계산되어, 턴 손익이 KIS
종목별 ROI와 정합한다. (이전 턴이 만든 미실현이 실현되는 턴에 온전히 귀속되므로, 구
MTM-리셋 설계가 지키던 '턴별 손익 합 = 누적 실현손익' 불변식은 더 이상 성립하지 않는다.)

이 모듈은 I/O를 하지 않는다. 원장 dict을 받아 제자리에서 갱신할 뿐이다.
"""

# Sim10의 국면 → 하위 전략 태그 (manifest의 실제 심 id)
REGIME_TAG = {
    'BULL': 'sim4_bull_daytrading',
    'SIDEWAYS': 'sim5_sideways',
    'BEAR': 'sim6',
}


def new_turn(turn_id: str, capital: float, positions: dict,
             opening_basis: dict | None = None, current_prices: dict | None = None) -> dict:
    """새 턴을 연다. 물려받은 보유 종목의 기준가를 **매입 평단(avg_price)**으로 잡는다.

    ON 시점 시세로 리셋(MTM)하지 않는다 — ON 전부터 보유한 종목도 원래 매입가부터 손익을
    재야 KIS 종목별 ROI와 턴 손익이 정합하기 때문이다. opening_basis/current_prices는 구
    스냅샷-리셋 방식의 잔재로, 호출부 호환을 위해 시그니처만 남기고 의도적으로 무시한다.
    """
    basis = {code: float(p.get('avg_price', 0)) for code, p in positions.items()}
    return {'id': turn_id, 'capital': float(capital), 'basis': basis,
            'by_tag': {}, 'active_tag': None}


def switch_tag(turn: dict, positions: dict, new_tag: str, current_prices: dict) -> None:
    """활성 전략을 전환한다. 직전 태그가 자기 구간의 평가손익을 확정 귀속받고, 기준가가 리셋된다.

    턴 첫 실행(active_tag=None)은 락인할 직전 태그가 없으므로 기준가를 그대로 둔다 —
    ON 시점부터 첫 실행까지의 변동은 첫 태그의 몫이다.
    """
    old = turn.get('active_tag')
    if old == new_tag:
        return
    basis = turn.setdefault('basis', {})
    by_tag = turn.setdefault('by_tag', {})
    for code, p in positions.items():
        if old is None:
            p['tag'] = new_tag           # 기준가는 new_turn이 잡은 ON 시점 값 유지
            continue
        px = float(current_prices.get(code) or 0)
        if px <= 0:
            continue                     # 시세 없음 → 락인 불가. 이 종목은 직전 태그·기준가 유지.
        b = float(basis.get(code, px))
        by_tag[old] = round(by_tag.get(old, 0.0) + (px - b) * p['quantity'], 2)
        basis[code] = px
        p['tag'] = new_tag
    turn['active_tag'] = new_tag


def record_buy(turn: dict, code: str, qty: int, price: float, prev_qty: int) -> None:
    """매수 체결 반영. 추가 매수 시 기준가는 평단처럼 가중평균된다.

    prev_qty는 체결 반영 '이전'의 보유 수량이다(호출부가 _apply_order_to_positions 전에 캡처).
    """
    basis = turn.setdefault('basis', {})
    price = float(price)
    if prev_qty > 0 and code in basis:
        basis[code] = (basis[code] * prev_qty + price * qty) / (prev_qty + qty)
    else:
        basis[code] = price


def record_sell(turn: dict, positions: dict, code: str, qty: int, price: float) -> None:
    """매도 체결 반영. 평단이 아니라 '기준가' 대비 손익을 그 종목을 들고 있던 태그에 귀속한다.

    활성 태그가 아니라 positions[code]['tag']를 쓰는 이유: switch_tag는 시세를 못 구한 종목을
    의도적으로 직전 태그·기준가에 남겨둔다. 그런 종목의 매도를 활성 태그에 귀속하면 기준가는
    옛것인데 손익은 새 태그가 가져가 SIM별 분해가 엇갈린다. 표시 계산(TS computeTurnPnl)도
    미실현을 pos.tag에 귀속하므로 양쪽 기준을 일치시킨다.
    """
    basis = turn.get('basis', {})
    if code not in basis:
        return
    tag = (positions.get(code) or {}).get('tag') or turn.get('active_tag')
    if not tag:
        return
    by_tag = turn.setdefault('by_tag', {})
    by_tag[tag] = round(by_tag.get(tag, 0.0) + (float(price) - basis[code]) * qty, 2)


def prune_basis(turn: dict, positions: dict) -> None:
    """전량 매도된 종목의 기준가를 정리한다(손익은 record_sell이 이미 귀속시켰다)."""
    basis = turn.get('basis', {})
    for code in list(basis):
        if code not in positions:
            basis.pop(code, None)
